src/externs.ts

```diff
@@ -0,0 +1,439 @@
+ *   /.. \@type {string} ./
+ *   Foo.prototype.bar;
+ *
+ * The generated externs indicate to Closure Compiler that symbols are extern to the optimization
+ * process, i.e. they are provided by outside code. That most importantly means they must not be
+ * renamed or removed.
+ */
+
+import * as path from 'path';
+
+import * as jsdoc from './jsdoc';
+import {AnnotatorHost, escapeForComment, isValidClosurePropertyName, maybeAddHeritageClauses} from './jsdoc_transformer';
+import {ModuleTypeTranslator} from './module_type_translator';
+import {getEntityNameText, getIdentifierText, hasModifierFlag, isDtsFileName} from './transformer_util';
+import * as ts from './typescript';
+
+/**
+ * Symbols that are already declared as externs in Closure, that should
+ * be avoided by tsickle's "declare ..." => externs.js conversion.
+ */
+const CLOSURE_EXTERNS_BLACKLIST: ReadonlyArray<string> = [
+  'exports',
+  'global',
+  'module',
+  // ErrorConstructor is the interface of the Error object itself.
+  // tsickle detects that this is part of the TypeScript standard library
+  // and assumes it's part of the Closure standard library, but this
+  // assumption is wrong for ErrorConstructor.  To properly handle this
+  // we'd somehow need to map methods defined on the ErrorConstructor
+  // interface into properties on Closure's Error object, but for now it's
+  // simpler to just blacklist it.
+  'ErrorConstructor',
+  'Symbol',
+  'WorkerGlobalScope',
+];
+
+/**
+ * generateExterns generates extern definitions for all ambient declarations in the given source
+ * file. It returns a string representation of the Closure JavaScript, not including the externs
+ * header.
+ */
+export function generateExterns(
+    typeChecker: ts.TypeChecker, sourceFile: ts.SourceFile,
+    host: AnnotatorHost): {output: string, diagnostics: ts.Diagnostic[]} {
+  let output = '';
+  const diagnostics: ts.Diagnostic[] = [];
+  const isDts = isDtsFileName(sourceFile.fileName);
+  const mtt =
+      new ModuleTypeTranslator(sourceFile, typeChecker, host, diagnostics, /*isForExterns*/ true);
+
+  for (const stmt of sourceFile.statements) {
+    if (!isDts && !hasModifierFlag(stmt, ts.ModifierFlags.Ambient)) continue;
+    visitor(stmt, []);
+  }
+
+  return {output, diagnostics};
+
+  function emit(str: string) {
+    output += str;
+  }
+
+  /**
+   * isFirstDeclaration returns true if decl is the first declaration
+   * of its symbol.  E.g. imagine
+   *   interface Foo { x: number; }
+   *   interface Foo { y: number; }
+   * we only want to emit the "\@record" for Foo on the first one.
+   */
+  function isFirstDeclaration(decl: ts.DeclarationStatement): boolean {
+    if (!decl.name) return true;
+    const sym = typeChecker.getSymbolAtLocation(decl.name)!;
+    if (!sym.declarations || sym.declarations.length < 2) return true;
+    return decl === sym.declarations[0];
+  }
+
+  function writeVariable(name: string, namespace: ReadonlyArray<string>, value?: string) {
```

Comment:
Docs on this and the one below it.  I think it might be better name 'writeVariableStatement' since it incldues the 'var' and the semi.

Reply:
Hm, I looked into this more and I guess this is just moved from the old code.  Still might be worth doing some minor fixes.